forc-pkg/src/source/git/auth.rs

```diff
@@ -0,0 +1,61 @@
+/// Handler holds all required information for handling authentication callbacks from `git2`.
+pub(crate) struct AuthHandler {
+    config: git2::Config,
+    /// Shows if the `AuthHandler` tried to make `SSH` authentication so far.
+    ssh_authentication_attempt: bool,
+    /// Shows if the `AuthHandler` tried to make `USER_PASS_PLAINTEXT` authentication so far.
+    plain_user_pass_attempt: bool,
+}
+
+impl AuthHandler {
+    /// Creates a new `AuthHandler` from all fields of the struct. If there are no specific reasons
+    /// not to, `default_with_config` should be prefered.
+    pub(crate) fn new(
+        config: git2::Config,
+        ssh_authentication_attempt: bool,
+        plain_user_pass_attempt: bool,
+    ) -> Self {
+        Self {
+            config,
+            ssh_authentication_attempt,
+            plain_user_pass_attempt,
+        }
+    }
+
+    /// Creates a new `AuthContext` with provided `git2::Config` and default values for other
+    /// context used during handling authentication callbacks.
+    pub(crate) fn default_with_config(config: git2::Config) -> Self {
+        let ssh_authentication_attempt = false;
+        let plain_user_pass_attempt = false;
+        Self::new(config, ssh_authentication_attempt, plain_user_pass_attempt)
+    }
+
+    pub fn handle_callback(
```

Comment:
Could we maybe add some docs here about when the given callback function will be called? Judging by the use of the `ssh_authentication_attempt` and `plain_user_pass_attempt` `bool`s it looks like this can get called multiple times - is that right?

Reply:
Yeah, the the remote callback that we register to `FetchOpts` calls this handler until it receives a valid authentication method or an explicit error. So without the flags in place users can end-up in infinite loops waiting to be obtain a sucessful SSH authentication